### PR TITLE
add directory provider and i22 example

### PR DESCRIPTION
**Pull request by rosesyrett**

### CODE REVIEW

**Review by keithralphs — COMMENTED** [review]
to be discussed further...

File: src/dodal/beamlines/i22.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        self.setpoint = epics_signal_rw(float, prefix + "RAMP:LIMIT:SET")
+        self.readback = epics_signal_r(float, prefix + "TEMP")
+        self.ramp_rate = epics_signal_rw(float, prefix + "RAMP:RATE:SET")
+        self.set_readable_signals(config=[self.ramp_rate], read=[self.readback])
+        super().__init__(name)
+
+    def set_name(self, name: str):
+        super().set_name(name)
+        # Readback should be named the same as its parent in read()
+        self.readback.set_name(name)
```
[the remark this answers is not in the document]
Comment (DiamondJoseph) [reply src/dodal/beamlines/i22.py]: Done

**Comment by DiamondJoseph** [comment]
Requires https://github.com/bluesky/ophyd-async/pull/24 and a release of that project
Requires a release of Bluesky with new event model changes

File: src/dodal/parameters/gda_directory_provider.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    def __init__(self, url: str, visit_id: str) -> None:
+        """url to the GDA visit service rest root endpoint"""
+        self.url = url
+        self.visit_id = visit_id
+        
+        try:
+            asyncio.run(asyncio.wait_for(self.connect("visits", response_type=InstrumentVisit, params={"visitId": self.visit_id}), timeout=15))
+        except TimeoutError:
+            raise Exception(f"Timeout trying to contact {self.url}/visits. Is GDA running?")
+
```
[the remark this answers is not in the document]
Comment (rosesyrett) [reply src/dodal/parameters/gda_directory_provider.py]: ^ i.e. don't raise an exception or do a try/catch here.

File: src/dodal/parameters/gda_directory_provider.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+class DataCollectionIdentifier(BaseModel):
+    collectionId: int
+
+
+class VisitDirectoryProviderConfig(BaseModel, frozen=True):
+    url: str
+    beamline: str
+    base_path: str
```
Comment (DiamondJoseph) [inline src/dodal/parameters/gda_directory_provider.py:16]: VDPC then becomes part of BlueAPI config and is passed down as part of make_directory_provider when constructing the singleton directory provider.

**Comment by DiamondJoseph** [comment]
TODO: Merge #198, #197  into this change

File: src/dodal/devices/linkam.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+        self.setpoint = epics_signal_rw(float, prefix + "RAMP:LIMIT:SET")
+        self.readback = epics_signal_r(float, prefix + "TEMP")
+        self.ramp_rate = epics_signal_rw(float, prefix + "RAMP:RATE:SET")
+        self.set_readable_signals(config=[self.ramp_rate], read=[self.readback])
+        super().__init__(name)
+
+    def set_name(self, name: str):
+        super().set_name(name)
+        # Readback should be named the same as its parent in read()
+        self.readback.set_name(name)
```
Comment (DiamondJoseph) [inline src/dodal/devices/linkam.py:16]: Is this superseded by linkam3.py?

Comment (rosesyrett) [reply src/dodal/devices/linkam.py:16]: Yes;we can remove this

**Comment by DominicOram** [comment]
@DiamondJoseph and @callumforrester, do we still need this PR? It's very stale...

**Comment by DiamondJoseph** [comment]
The only work in this patchset that isn't already merged into this/another repository is the Linkam.

**Comment by callumforrester** [comment]
Indeed, would be nice to have it back but it's not a priority

**Comment by DominicOram** [comment]
I mean feel free to open a PR with it that's juts linkam, I don't mind reviewing it. My concern was this PR is old and very broad. Meaning that the path to merging it is not clear

**Comment by callumforrester** [comment]
Will do, but as I say it's not a priority and we already have a few PRs backed up behind the `ophyd-async` issues (#533)

**Pull request closed without merging**